src/dune_rules/coq_lib.ml

```diff
@@ -38,128 +67,333 @@ let libraries l = l.libraries
+                  with
+                  | false -> Resolve.return ()
+                  | true ->
+                    Resolve.fail
+                    @@ User_message.make ~loc
+                         [ Pp.textf
+                             "private theory %s may not depend on a public \
+                              library"
+                             (Coq_lib_name.to_string (snd s.name))
+                         ]
+              in
+              (loc, lib))
+        and+ theories =
+          let check_boot (lib : lib) =
+            let open Resolve.O in
+            let* boot = boot in
+            match boot with
+            | None -> Resolve.return ()
+            | Some boot -> (
+              let* boot' = lib.boot in
+              match boot' with
+              | None -> Resolve.return ()
+              | Some boot' -> (
+                match Id.compare boot.id boot'.id with
+                | Eq -> Resolve.return ()
+                | _ -> Error.incompatible_boot lib.id id))
+          in
+          Resolve.Memo.List.map s.buildable.theories ~f:(fun (loc, theory) ->
+              let open Resolve.Memo.O in
+              let* theory = resolve ~coq_lang_version coq_db (loc, theory) in
+              let* () = Resolve.Memo.lift @@ check_boot theory in
+              let+ () =
+                if allow_private_deps then Resolve.Memo.return ()
+                else
+                  match theory.package with
+                  | Some _ -> Resolve.Memo.return ()
+                  | None ->
+                    Error.private_deps_not_allowed ~loc (snd theory.name)
+              in
+              (loc, theory))
+        in
+        let theories =
+          let open Resolve.O in
+          let* boot = boot in
+          match boot with
+          | None -> theories
+          | Some boot ->
+            let+ theories = theories in
+            (* TODO: if lib is boot, don't add boot dep *)
+            (* maybe use the loc for boot? *)
+            (Loc.none, boot) :: theories
+        in
+        let map_error x =
+          let human_readable_description () = Id.pp id in
+          Resolve.push_stack_frame ~human_readable_description x
+        in
+        let theories = map_error theories in
+        let libraries = map_error libraries in
+
+        { name = s.name
+        ; boot
+        ; id
+        ; implicit = s.boot
+        ; obj_root = dir
+        ; src_root = dir
+        ; theories
+        ; libraries
+        ; theories_closure =
+            lazy
+              (Resolve.bind theories ~f:(fun theories ->
+                   List.map theories ~f:snd |> top_closure))
+        ; package = s.package
+        }
+      in
+      let module Input = struct
+        type nonrec t = t * Lib.DB.t * Path.Build.t * Coq_stanza.Theory.t
+
+        let equal (coq_db, ml_db, path, stanza) (coq_db', ml_db', path', stanza')
+            =
+          coq_db == coq_db' && ml_db == ml_db'
```

Comment:
This physical equality use looks a bit scary, doesn't it?

Reply:
In this case, all it's used for is for detect if the inputs have changed and recomputation is necessary. So it's actually the most conservative option and can only cause unnecessary recomputation of the memoized function.

Reply:
Yeah, but if that would happen, it would be pretty bad, right? After all the whole point of Dune is to avoid recomputing stuff, isn't it?

Moreover this seems very brittle, as we don't have tests for this, so even if it works properly today, this pattern doesn't seem very stable w.r.t. refactoring.

Reply:
We can define `Lib.DB.equal` and `Coq_lib.DB.equal` as physical equality. At least that would keep things behind an interface. But I'm afraid that there's no better equality we could select for databases. Our databases are lazy loaded, so it doesn't make sense to compare them structurally.

> After all the whole point of Dune is to avoid recomputing stuff, isn't it?

The worst case for recomputation here would be once per watch mode run. Still totally acceptable (it's what we have right now anyway due to other reasons).

> Moreover this seems very brittle, as we don't have tests for this, so even if it works properly today, this pattern doesn't seem very stable w.r.t. refactoring.

It's a bit of an open problem in dune currently. We don't currently have a good idea what gets recomputed and how often. As I mentioned, in this case, our behavior is strictly no worse than before, but it would be hard to measure if we improved things. I'd go for better benchmarks and reserve the unit tests for correctness here personally though.

Reply:
Indeed, a big problem with `Memo.t` is that hashing can quickly become dominant if you have a structural representation, and if you don't, you can't do it anyways. I am not sure I like how the actual database and the resolution mechanism are coupled in this design, but I need to think a bit more about it tho.

Reply:
It's basically the same design we already have for OCaml libraries. If you think you can simplify it somehow, do let us know.

But before dwelling on a new design, let's make sure the current one suffers from performance problems first :) After all, there are no issues with correctness here. Would be nice to see if library resolution is something that is indeed slowing something like the coq-universe in watch mode.

Reply:
@rgrinberg indeed here what makes me think is more a matter of implicit vs explicit representation. For example, with this design, asking Dune (for UI use for example) "what theories are in scope in this build" seems a bit difficult, isn't it?

Similar questions do arise for example with the representation of the dep DAG, some people may find it useful having an explicit representation for example for research purposes.

But indeed, I've been aware of this design since a long time, and I need to think about it.

Reply:
> "what theories are in scope in this build" seems a bit difficult, isn't it?

We handle that separately that case as well on the OCaml side. See `Lib.DB.all` and the `all` field in `Lib.DB.t`.

> Similar questions do arise for example with the representation of the dep DAG, some people may find it useful having an explicit representation for example for research purposes.

Keep in mind that `Coq_lib.DB.t` will need to handle two types of db's:

* Constructed by traversing stanzas in a scope
* Lazily loaded from `user-contrib`.

The former would allow an explicit representation (though you could still speed it up a bunch with memoization), but the latter will not.

Reply:
Can't the explicit representation be build from scanning (lazily) lib?

Reply:
I'd need to understand how "explicit" you want to go, but I do think that it would necessarily need to be more eager than we might want if we support multiple directories in `COQPATH`.

Reply:
For now I'm fine with the current state of affairs, sorry if I was going out tangent with this.

COQPATH support is a different issue, we need indeed to have a lazy scanning system for `installed_theories` that indeed can merge the theory three.

That's not conceptually complicated, but annoying!

So when resolving an `installed_theory`, indeed we first split the theory in its path components, then we need to check the directory exists in some of the set `user-contrib:$COQPATH`, then do that recursively for the second component.

It there is a conflict, we error (as coqc should do, but not in the Coq test suite) (@Alizter that'd be a nice addition actually to Coq's test-suite)